libcxx/include/__algorithm/radix_sort.h

```diff
@@ -0,0 +1,427 @@
+}
+
+template <typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _Map, typename _Radix>
+_LIBCPP_HIDE_FROM_ABI typename enable_if<
+    __radix_sort_traits<typename iterator_traits<_RandomAccessIterator1>::value_type, _Map, _Radix>::radix_count % 2 ==
+        0,
+    void>::type
+__radix_sort_impl(_RandomAccessIterator1 __first,
+                  _RandomAccessIterator1 __last,
+                  _RandomAccessIterator2 __buffer_begin,
+                  _Map __map,
+                  _Radix __radix) {
+  using value_type = typename iterator_traits<_RandomAccessIterator1>::value_type;
+  using traits     = __radix_sort_traits<value_type, _Map, _Radix>;
+
+  using difference_type = typename iterator_traits<_RandomAccessIterator1>::difference_type;
+  difference_type __counters[traits::radix_count][traits::radix_value_range] = {{0}};
+  difference_type __maximums[traits::radix_count]                            = {0};
+  const auto __is_sorted = std::__collect(__first, __last, __map, __radix, __counters, __maximums);
+  if (not __is_sorted) {
+    const auto __range_size = std::distance(__first, __last);
+    auto __buffer_end       = __buffer_begin + __range_size;
+    for (size_t __radix_number = 0; __radix_number < traits::radix_count; __radix_number += 2) {
+      const auto __n0th_is_single = __maximums[__radix_number] == __range_size;
+      const auto __n1th_is_single = __maximums[__radix_number + 1] == __range_size;
+
+      if (__n0th_is_single && __n1th_is_single) {
+        continue;
+      }
+
+      if (__n0th_is_single) {
+        std::copy(std::__move_assign_please(__first), std::__move_assign_please(__last), __buffer_begin);
+      } else {
+        auto __n0th = [__radix_number, &__map, &__radix](const auto& __v) {
+          return std::__nth_radix(__radix_number, __radix)(__map(__v));
+        };
+        std::__dispose_backward(
+            std::__move_assign_please(__first),
+            std::__move_assign_please(__last),
+            __buffer_begin,
+            __n0th,
+            __counters[__radix_number]);
+      }
+
+      if (__n1th_is_single) {
+        std::copy(std::__move_assign_please(__buffer_begin), std::__move_assign_please(__buffer_end), __first);
+      } else {
+        auto __n1th = [__radix_number, &__map, &__radix](const auto& __v) {
+          return std::__nth_radix(__radix_number + 1, __radix)(__map(__v));
+        };
+        std::__dispose_backward(
+            std::__move_assign_please(__buffer_begin),
+            std::__move_assign_please(__buffer_end),
+            __first,
+            __n1th,
+            __counters[__radix_number + 1]);
+      }
+    }
+  }
+}
+
+_LIBCPP_HIDE_FROM_ABI constexpr auto __to_unsigned(bool __b) { return __b; }
+
+template <typename _Ip>
+_LIBCPP_HIDE_FROM_ABI constexpr auto __to_unsigned(_Ip __n) {
+  constexpr const auto __min_value = numeric_limits<_Ip>::min();
+  return static_cast<make_unsigned_t<_Ip> >(__n ^ __min_value);
+}
```

Comment:
I think we can't. As I can see, `__to_unsigned_like` is a static cast to corresponding unsigned. But it's not just enough to cast to unsigned. We need to shift the value range from `[0, 2^n)` to `[-2^(n-1), 2^(n-1))`, so that if `x` is less than `y` in the old range, then `x` will also be less than `y` in the new range.

Reply:
> We need to shift the value range from `[0, 2^n)` to `[-2^(n-1), 2^(n-1))`, so that if `x` is less than `y` in the old range, then `x` will also be less than `y` in the new range.

Hmm, I think this function actually shifts from `[-2^(n-1), 2^(n-1))` to `[0, 2^n)`, which should be intended. How about changing the name to `__shift_to_unsiged` or `__plus_to_unsigned` as a plain `to` has closer meaning to casting in the standard library?

Reply:
Yes, you're right, I wanted to write that values are shifted from `[-2^(n-1), 2^(n-1))` to `[0, 2^n)`.
`__shift_to_unsiged` is a good name, thanks. Changed it.